tests/test_solvers_perturbed_maximization.py

```diff
@@ -0,0 +1,688 @@
+    )
+    check_test_solution(solver, 100, 0.2)
+
+
+def test_alternates():
+    """Test that alternates are selected correctly"""
+
+    # probability limits 1 case
+    S = np.transpose(np.array([[0, 0.3, 1], [1, 0.5, 0.4], [0.4, 1, 0.3]]))
+    M = np.zeros(np.shape(S))
+    Q = np.full(np.shape(S), 1.0)
+
+    solver = PerturbedMaximizationSolver(
+        [0, 0, 0], [1, 1, 1], [1, 1, 1], encoder(-S, M, Q, 0.0)
+    )
+    solution = np.transpose(np.array([[0, 0, 1], [1, 0, 0], [0, 1, 0]]))
+    solver.solve()
+    answer = solver.get_alternates(2)
+    alternates = {0: [2, 0], 1: [1, 0], 2: [1, 2]}
+    alt_probs = np.ones((3, 3)) - solution
+    assert np.all(np.isclose(solver.fractional_assignment_matrix, solution))
+    assert np.all(np.isclose(solver.alternate_probability_matrix, alt_probs))
+    assert np.all(answer == alternates)
+
+    # test with lower probability limits
+    Q = np.full(np.shape(S), 0.7)
+    solver = PerturbedMaximizationSolver(
+        [0, 0, 0], [1, 1, 1], [1, 1, 1], encoder(-S, M, Q, 0.0)
+    )
+    solution = np.transpose(
+        np.array([[0, 0.3, 0.7], [0.7, 0, 0.3], [0.3, 0.7, 0]])
+    )
+    alt_probs = np.transpose(
+        np.array([[0.7, 4 / 7, 0], [0, 0.7, 4 / 7], [4 / 7, 0, 0.7]])
+    )
+    solver.solve()
+    solver.get_alternates(2)
+    assert np.all(np.isclose(solution, solver.fractional_assignment_matrix))
+    assert np.all(np.isclose(alt_probs, solver.alternate_probability_matrix))
+
+
+def test_opt_fraction():
+    """Test that fraction of opt is calculated correctly"""
+    S = np.eye(5)
+    M = np.zeros(np.shape(S))
+    Q = np.full(np.shape(S), 0.5)
+    solver = PerturbedMaximizationSolver(
+        [0, 0, 0, 0, 0],
+        [1, 1, 1, 1, 1],
+        [1, 1, 1, 1, 1],
+        encoder(-S, M, Q, 0.0),
+        True,  # allow zero score assignment
+    )
+
+    solver.solve()
+    assert solver.solved and solver.deterministic_assignment_solved
+    assert np.isclose(solver.fractional_assignment_cost, -2.5)
+    assert np.isclose(solver.deterministic_assignment_cost, -5)
+    assert np.isclose(solver.get_fraction_of_opt(), 0.5)
+
+    S = np.eye(5)
+    for i in range(1, 5):
+        S[i - 1, i] = 1
+    S[4, 0] = 1
+    solver = PerturbedMaximizationSolver(
+        [0, 0, 0, 0, 0],
+        [1, 1, 1, 1, 1],
+        [1, 1, 1, 1, 1],
+        encoder(-S, M, Q, 0.0),
+        True,  # allow zero score assignment
+    )
+
+    solver.solve()
+    assert solver.solved and solver.deterministic_assignment_solved
+    assert np.isclose(solver.fractional_assignment_cost, -5)
+    assert np.isclose(solver.deterministic_assignment_cost, -5)
+    assert np.isclose(solver.get_fraction_of_opt(), 1)
+
+    Q = np.full(np.shape(S), 0.3)
+    solver = PerturbedMaximizationSolver(
+        [0, 0, 0, 0, 0],
+        [1, 1, 1, 1, 1],
+        [1, 1, 1, 1, 1],
+        encoder(-S, M, Q, 0.0),
+        True,  # allow zero score assignment
```

Comment:
could you add a test here allow_zero_score_assignments = No and all the scores are zero? we should get a no solution error in this case.

Reply:
Done. But in that case, instead of throwing an exception, I let solver.solved = False and return.
This is also done in the RandomizedSolver. I am trying to keep it consistent.